src/cdomains/unionDomain.ml

```diff
@@ -29,21 +29,22 @@ struct
     | Cil.NoOffset ->
       let c_lval = Option.get lval in
       begin match lift_f with
-      | `Lifted f ->
-        let f_lval = Cil.addOffsetLval (Field (f, NoOffset)) c_lval in
-        value_invariant ~offset ~lval:(Some f_lval) v
-      | `Top
-      | `Bot ->
-        Invariant.none
+        | `Lifted f ->
+          let f_lval = Cil.addOffsetLval (Field (f, NoOffset)) c_lval in
+          value_invariant ~offset ~lval:(Some f_lval) v
+        | `Top
+        | `Bot ->
+          Invariant.none
       end
     (* invariant for one field *)
     | Field (f, offset) ->
-      let c_lval = Option.get lval in
       begin match lift_f with
         | `Lifted f' ->
           let v = Values.cast ~torg:f'.ftype f.ftype v in
-          let f_lval = Cil.addOffsetLval (Field (f, NoOffset)) c_lval in
-          value_invariant ~offset ~lval:(Some f_lval) v
+          (* Do not add the field offset to lval here:
+             Otherwise, for pointers to a member type of the union, this would
+             generate an assertion with illegal field offset. *)
```

Comment:
What's the example/test case for this? The pointer to an union field should be equal to the pointer to the union itself.
It's also out of sync with the `NoOffset` case now.

Reply:
>What's the example/test case for this?

I added test case [`61/01`](https://github.com/goblint/analyzer/blob/fix_evalAssert/tests/regression/61-evalAssert/01-union_evalAssert.c).
https://github.com/goblint/analyzer/blob/a4f0a6e8749227e6eabb8ae0f4af155a95721f5a/tests/regression/61-evalAssert/01-union_evalAssert.c#L4-L23

 If one does not change the implementation here, then there will be invariants generated about `ptr` such as 
`assert(ptr->str.a == 1);`
`gcc` will complain that the `struct s` has no member `str`. 
The fix though is still flawed, as `bench/coreutils/dd_comb.c` shows: There I had to [change some asserts about unions](https://gitlab.com/erhardj/sv-benchmarks/-/commit/d10a2fe9ab4ed29d0d3c5f4258240fbbd7136f60#714c4d12a37819c79ac9db33e779ebd886327525_87516_87513), because the offset was missing.

>The pointer to an union field should be equal to the pointer to the union itself.

When one has a pointer to a union field, then generating the assertion should not add this offset again to the pointer (see example).

Reply:
#861 works on the example, and does not create the problems on `dd_comb.c`.